ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/LedgerSeq.hs

```diff
@@ -0,0 +1,485 @@
+-- the @HardForkState@'ed state.
+--
+-- Therefore it sounds reasonable to hold a @LedgerState blk EmptyMK@ with no
+-- values, and a @LedgerTables blk ValuesMK@ next to it, that will live its
+-- entire lifetime as @LedgerTables@ of the @HardForkBlock@.
+data StateRef m l = StateRef {
+    state  :: !(l EmptyMK)
+  , tables :: !(LedgerTablesHandle m l)
+  } deriving (Generic)
+
+deriving instance (IOLike m, NoThunks (l EmptyMK)) => NoThunks (StateRef m l)
+
+instance Eq (l EmptyMK) => Eq (StateRef m l) where
+  (==) = (==) `on` state
+
+instance Show (l EmptyMK) => Show (StateRef m l) where
+  show = show . state
+
+instance GetTip l => Anchorable (WithOrigin SlotNo) (StateRef m l) (StateRef m l) where
+  asAnchor = id
+  getAnchorMeasure _ = getTipSlot . state
+
+{-------------------------------------------------------------------------------
+  The LedgerSeq
+-------------------------------------------------------------------------------}
+
+newtype LedgerSeq m l = LedgerSeq {
+    getLedgerSeq :: AnchoredSeq (WithOrigin SlotNo) (StateRef m l) (StateRef m l)
+  } deriving (Generic)
+
+deriving newtype instance (IOLike m, NoThunks (l EmptyMK)) => NoThunks (LedgerSeq m l)
+
+deriving newtype instance Eq   (l EmptyMK) => Eq   (LedgerSeq m l)
+deriving newtype instance Show (l EmptyMK) => Show (LedgerSeq m l)
+
+type LedgerSeq' m blk = LedgerSeq m (ExtLedgerState blk)
+
+{-------------------------------------------------------------------------------
+  Construction
+-------------------------------------------------------------------------------}
+
+-- | Creates an empty @LedgerSeq@
+empty ::
+     ( GetTip l
+     , IOLike m
+     )
+  => l EmptyMK
+  -> LedgerTables l ValuesMK
+  -> (LedgerTables l ValuesMK -> m ( LedgerTablesHandle m l))
+  -> m (LedgerSeq m l)
+empty st tbs new = LedgerSeq . AS.Empty . StateRef st <$> new tbs
+
+-- | Creates an empty @LedgerSeq@
+empty' ::
+     ( GetTip l
+     , IOLike m
+     , HasLedgerTables l
+     )
+  => l ValuesMK
+  -> (LedgerTables l ValuesMK -> m (LedgerTablesHandle m l))
+  -> m (LedgerSeq m l)
+empty' st = empty (forgetLedgerTables st) (ltprj st)
+
+closeLedgerSeq :: Monad m => LedgerSeq m l -> m ()
+closeLedgerSeq = mapM_ (close . tables) . toOldestFirst . getLedgerSeq
+
+{-------------------------------------------------------------------------------
+  Apply blocks
+-------------------------------------------------------------------------------}
+
+-- | If applying a block on top of the ledger state at the tip is succesful,
+-- extend the DbChangelog with the resulting ledger state.
+--
+-- Note that we require @c@ (from the particular choice of @Ap m l blk c@) so
+-- this sometimes can throw ledger errors.
+reapplyThenPush :: (IOLike m, ApplyBlock l blk)
+              => ResourceRegistry m
+              -> LedgerDbCfg l
+              -> blk
+              ->    LedgerSeq m l
+              -> m (LedgerSeq m l, LedgerSeq m l)
```

Comment:
The Haddock should explain why there are two `LedgerSeq`s returned, ie that the `fst` one should be `close`d.